src/nedap/speced/def.cljc

```diff
@@ -8,19 +8,20 @@
   (:require
    #?(:clj [clojure.test :as test])
    #?(:clj [nedap.speced.def.impl.fn :as impl.fn] :cljs [nedap.speced.def.impl.dummy :as impl.fn])
+   #?(:clj nedap.speced.def.impl.spec-assertion.jvm)
    #?(:clj [clojure.core.specs.alpha :as specs] :cljs [cljs.core.specs.alpha :as specs])
    #?(:clj [nedap.speced.def.impl.letfn :as impl.letfn])
    #?(:clj [nedap.speced.def.impl.let-impl :as let-impl])
+   #?(:clj [nedap.speced.def.impl.spec-assertion])
    [clojure.spec.alpha :as spec]
    [nedap.speced.def.doc :refer [doc-registry rebl-doc-registry]]
    [nedap.speced.def.impl.def-with-doc]
    [nedap.speced.def.impl.defn :as impl.defn]
    [nedap.speced.def.impl.defprotocol]
    [nedap.speced.def.impl.doc :as impl.doc]
    [nedap.speced.def.impl.satisfies]
-   [nedap.speced.def.impl.spec-assertion]
    [nedap.utils.spec.api :refer [check!]])
-  #?(:cljs (:require-macros [nedap.speced.def :refer [def-with-doc]])))
+  #?(:cljs (:require-macros [nedap.speced.def :refer [def-with-doc]] [nedap.speced.def.impl.spec-assertion.cljs])))
```

Comment:
to aid editors, we could use [the same trick as matcher-combinators](https://github.com/nubank/matcher-combinators/blob/2ae99ff8731027d122da6e8f01ba2f7fcfd9b74a/src/cljc/matcher_combinators/test.cljc#L9-L12).
e.g. declare `spec-assertion-thrown?` in this namespace so it can be referred.

Reply:
Sounds fair, although also I'm thinking of making `spec-assertion-thrown?` opt-in at some point. See readme:

> In a future, we might add a way to disable this side-effect.

By following the trick, we further solidify an "API" to could be broken relatively soon.

Reply:
That could still be done, right? Just have the side-effecting ns declare the var.

Reply:
mmm

> e.g. declare spec-assertion-thrown? in this namespace so it can be referred.

does this mean "refer at clj level" or "resolve a Cursive warning by tricking it into believe a certain var exists"?

Reply:
> "resolve a Cursive warning by tricking it into believe a certain var exists"?

essentially this; example: https://github.com/nedap/utils.test/blob/2413cd8da2ab58a626f9385e30d3598338af1856/test/unit/nedap/utils/test/api.cljc#L5

This example would work without the refer; as matcher-combinators is (probably also) loaded earlier. but to prevent the editor from complaining about an unresolvable var, you refer this. IMO it also makes it more tracible (at least you know it's added by matcher-combinator)

Reply:
I see what you mean / what is achieved, but also as I see it, the whole pattern is misleading.

In the linked ns, `(match? expected` is _not_ an invocation to `#'matcher-combinators.test/match?`. That's pretty misleading (because the :refer suggests otherwise), and has a pretty high "wat" factor for anyone who is not a Cursive user who works at Nedap.

Not sure if I'd like to reinforce that pattern, particularly in a library like `speced.def` which tries to be as agnostic as possible.

Reply:
>  That's pretty misleading

I see what you're saying. I must admit that the way `clojure.test/is`-assertions are expandable surprised me to say the least, more than once 😅; whether this makes it worse idk.

>  and has a pretty high "wat" factor for anyone who is not a Cursive user who works at Nedap.

It's not an internal practice, as matcher-combinators is not an internal library. I've seen it in an couple other places as well (https://github.com/metosin/testit#extending-testit-with-your-own-arrows,) I agree however that, it's not conventional per-se.

> which tries to be as agnostic as possible.

Keep in mind you're not required to use this; afaic it's determined by the consumer

Personally, as it aids tracability, i like it over the current impl.

Reply:
Delicate topic! I don't think there's even a right answer, particularly given `the way 'is' assertions are expandable surprised me to say the least`, which is hard to disagree with :)

Factually though, this seems a compensation over a specific IDE's intricacies. Where to draw the line?

* Should I e.g. adopt different development practices because a given linter cannot understand my macros?
  * What does that say about a runtime-oriented lang like Clojure? Maybe, that that facet is still not fully acknlowledged?
* Should I change `ex-info`->`js/Error` because of shadow-cljs intricacies? (https://github.com/nedap/speced.def/issues/100#issuecomment-570865803)
  * At first I thought _yes_, not anymore upon further research

I'd be happy to solve this somehow the day we address `In a future, we might add a way to disable this side-effect`. Would be nice to set the precedent of delivering `is` assertions more cleanly - requiring a lib (particularly transitively) shouln't magically expand `is` semantics.

Reply:
Thanks for being understanding with the topic.

As one last thought, I wonder if Cursive could use a kondo-style dotfile aiding var resolution etc.

Reply:
> As one last thought, I wonder if Cursive could use a kondo-style dotfile aiding var resolution etc.

For macro's etc. it's already possible, but the actual issue is multimethod support I think: https://github.com/cursive-ide/cursive/issues/238